### PR TITLE
[PyTorch] torch.compile for an OperationFuser group holding one operation

### PR DESCRIPTION
**Pull request by pggPL**
# Description

Add the infrastructure for `torch.compile(fullgraph=True)` on an `OperationFuser` group containing one basic operation. Dynamo traces the pipeline-level `_OperationFuserAutogradFunction`, while the operation can use independently registered forward and backward custom ops inside it. Existing forward and backward fusion plans remain separate.

This PR exercises the new interface with test operations. Converting production operations to it is follow-up work.

## Type of change

- [ ] Documentation change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Infra/Build change
- [x] Code refactoring

### Operation interface

`FusibleOperation` owns the shared compute interface for basic and fused operations:

- `fwd_args_type` and `bwd_args_type` are optional dataclass declarations defining each direction's custom-op input schema. Each declared direction is registered independently.
- `pack_forward_args(...)` and `pack_backward_args(...)` gather the fuser's inputs, module state and saved contexts into those dataclasses.
- `forward_compute(args)` returns `(output, extra_outputs_per_basic_op, aux)`.
- `forward_setup_context(basic_op_ctxs, args, aux)` explicitly saves the state needed by backward. Its default implementation does nothing.
- `backward_compute(args)` returns `(grad_input, grad_params_per_basic_op, grad_extra_inputs_per_basic_op)`. A `None` input gradient retains the existing pass-through convention.
- `forward_compute_fake(args)` and `backward_compute_fake(args)` describe the same results using `TensorSpec` without running the real computation.

`fuser_forward` and `fuser_backward` select direct compute or the registered custom op. `BasicOperation` also preserves its legacy `op_forward` / `op_backward` implementations. The compute bodies used as custom ops must not mutate inputs or return tensors aliasing their inputs; inputs and parameters needed by backward are saved through `forward_setup_context`.

`register_custom_op(op_name=..., arg_type=..., impl=..., fake_impl=...)` registers **one** autograd-free custom op and returns one callable, or `None` when the required registration APIs are unsupported. Nested tensor results and `None` leaves are preserved. Duplicate operator names raise `ValueError`. `register_custom_op_with_autograd` retains the autograd-wired interface used by `te.Linear`.

### Independent directions and compile limits

`compile_ops` is always a `(forward_callable, backward_callable)` tuple; either entry may be `None`. `compile_unsupported_reason(mode)` and the fuser's custom-op selection are evaluated separately for forward and backward. Backward eligibility is not checked when backward is unnecessary.

Custom-op execution remains limited to a single basic operation, with no fusion in the selected pass and no extra tensor channels. Explicit forward kwargs must be tensors, and quantizers must support value-opaque representation. Kwarg names are handled by `pack_forward_args`; there is no separate name declaration or inference mechanism.

When a direction cannot use a custom op, it uses its existing compute or legacy implementation. This does not guarantee that arbitrary legacy implementations support `fullgraph=True` or that backward can introduce a graph break. Directional fused implementations can use the shared interface, but automatic custom-op execution of fused plans remains outside this PR's scope.

### Saved state and tracing

- Copy operation contexts before modifying them in the traced backward.
- Avoid tensor-attribute mutation and tensor-storage cleanup while tracing.
- Save bare quantized storage as tensor buffers plus reconstruction metadata under compile.
- Reject delayed-scaling state under compile, including `CustomRecipe` with `DelayedScalingRequest`, rather than silently omitting its backward updates.

## Testing

The ops section of `test_torch_compile.py` covers 21 parametrized cases:

- Seven pipeline configurations in FP32 and BF16, including forward-only, backward-only and neither custom op. Tests compare outputs and input/parameter gradients with native PyTorch autograd and check custom-op presence separately in each graph. Forward-only also covers `no_grad()`.
- Four registration combinations, checking operator names and optional callables with a stub registrar.
- Direct and custom-op execution of a fused compute contract with two contexts, a residual input, an extra output, and nested results.
- Quantized tensor kwargs changing without recompilation, plus scalar-kwarg fallback.

Directional execution tests disable selected callables on the existing test operation; registration is tested separately. They do not claim coverage of arbitrary backward graph breaks.

Validation for head `83c95e56`:

- Focused workstation run: **21 passed**; Black and diff checks passed.
- B200 CI: `test_torch_compile.py` **177 passed, 22 skipped, 1 xpassed**; `test_fusible_ops.py` **2388 passed, 3623 skipped**.
- L0 passed on A100/H100/L40; L1 distributed passed on H100, along with MCore/FSDP and ONNX jobs. L0 B200 reported a strict XPASS in the selective-attention-recompute hybrid-quantization test: its assertions passed despite an expected-failure marker. The cause of that XPASS has not been established.

# Checklist:

- [x] I have read and followed the contributing guidelines
- [ ] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/ops/op.py [hunk 1]
```diff
@@ -537,15 +579,24 @@ def op_forward(
...
 
         """
+        return self._forward(
+            self.forward_impl,
+            ctx,
+            input_,
+            prev_op_grad_output_quantizer=prev_op_grad_output_quantizer,
+            next_op_input_quantizer=next_op_input_quantizer,
+            **kwargs,
+        )
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/op.py]: Be advised that `op_forward`/`op_backward` are convenience functions for simple operations. Anything that requires anything advanced (e.g. `GroupedLinear`) should override `fuser_forward`/`fuser_backward` instead, and I think `torch.compile` support falls into that category.

File: transformer_engine/pytorch/ops/op.py [hunk 2]
```diff
@@ -561,6 +612,118 @@ def op_backward(
...
+    def backward_impl(cls, args: Any) -> tuple:
+        """Backward over ``args``; returns grad_input, then parameter grads."""
+        raise NotImplementedError
+
+    @classmethod
+    def backward_fake(cls, args: Any) -> tuple:
+        """Shape-only twin of :meth:`backward_impl`."""
+        raise NotImplementedError
+
+    def resolve_fwd_args(
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/op.py]: I see this function is intended to pack the arguments for `forward_impl` into a dataclass. Can you explain why that is needed for `torch.compile`? I assume it's because the infrastructure in https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/dynamo/custom_op.py is constructing PyTorch ops that have a fixed function signature?

File: transformer_engine/pytorch/ops/op.py [hunk 3]
```diff
@@ -561,6 +612,118 @@ def op_backward(
...
+
+    def compiled_op_backward(
+        self,
+        ctx: OperationContext,
+        grad_output: torch.Tensor,
+    ) -> tuple[torch.Tensor, Iterable[Optional[torch.Tensor]]]:
+        """:meth:`op_backward` through this operation's custom op."""
+        return self._backward(self.compile_ops[1], ctx, grad_output)
+
+    def _forward(
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/op.py]: We're multiplying forward functions: `forward`, `_forward`, `forward_impl`, `op_forward`, `fuser_forward`, `compiled_op_forward`. I don't find it obvious what each of these is doing.

If I understand correctly, what we really want is to have separate stages for forward compute and context saving. This makes me think our design should look something like:
```python
# Pack args from OperationFuser into dataclass
def forward_pack_args(...):
    ...

# Actually do the forward compute
def forward_compute(args):
    ...

# Cache state state for backward
def forward_setup_context(ctx, args, aux):
    ...

# Simple forward implementation (not always implemented)
def op_forward(...):
    ...

# Advanced forward implementation
def fuser_forward(...):
    if op_forward is implemented:
        return output = self.op_forward(...), ...

    # If fuser_forward has not been overridden, assume that the torch.compile-compatible APIs have been defined
    args = self.pack_forward_args(...)
    output, aux = self.forward_compute(args)
    if ctx.requires_grad:
        self.forward_setup_context(ctx, args, aux)
    return output, ...
    
# torch.compile interface
def compiled_op_forward(...):
    args = self.pack_forward_args(...)
    output, aux = self.compile_ops[0](args)
    if ctx.requires_grad:
        self.forward_setup_context(ctx, args, aux)
    return output, ...
```

We should also apply a similar design for the backward functions.

File: transformer_engine/pytorch/ops/op.py [hunk 4]
```diff
@@ -186,6 +188,45 @@ class BasicOperation(FusibleOperation, metaclass=abc.ABCMeta):
...
+        )
+        cls.compile_ops = register_custom_op(
+            op_name=cls.__name__.lower(),
+            fwd_arg_type=cls.fwd_args_type,
+            fwd_impl=cls.forward_impl,
+            fwd_fake_impl=cls.forward_fake,
+            bwd_arg_type=cls.bwd_args_type,
+            bwd_impl=cls.backward_impl,
+            bwd_fake_impl=cls.backward_fake,
+        )
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/op.py]: This makes sense when we're dealing with basic ops. When we extend to support fused ops, can we register forward-only or backward-only PyTorch ops?

File: transformer_engine/pytorch/dynamo/custom_op.py [hunk 5]
```diff
@@ -1166,7 +1174,238 @@ def _all_quantized_tensor_subclasses() -> List[type]:
...
+    out_plan = _OutputPlan.parse(fwd_fake_impl(spec_obj))
+    return out_plan, fwd_op(fwd_args)
+
+
+# --------------------------------------------------------------------------- #
+# Op registration: the autograd-free pair, and the autograd-wired variant
+# --------------------------------------------------------------------------- #
+
+
 def register_custom_op(
```
Comment (timmoon10) [inline transformer_engine/pytorch/dynamo/custom_op.py:1252]: I see this function assumes both the forward and backward impls are available. When we do add support for forward-only or backward-only custom ops, it will just be a matter of skipping registration when `fwd_impl=None` or `bwd_impl=None`?

File: transformer_engine/pytorch/ops/op.py (hunk 2 above)
Comment (pggPL) [reply transformer_engine/pytorch/ops/op.py]: Torch.compile needs to know the type of every argument for forward/backward and also some information which tensors need gradient etc. - that was motivation for introducing this objects in Linear. Also, I want to avoid using parameters of BasicOperation or global state inside the ops - they should be resolved in resolve_fwd_args and passed as an arguments which will enable torch.compile to set proper guards and know when to recompile.

File: transformer_engine/pytorch/dynamo/custom_op.py (hunk 5 above)
Comment (pggPL) [reply transformer_engine/pytorch/dynamo/custom_op.py:1252]: good catch

File: transformer_engine/pytorch/ops/op.py (hunk 3 above)
Comment (pggPL) [reply transformer_engine/pytorch/ops/op.py]: That makes sense, I made appropriate changes.

File: transformer_engine/pytorch/ops/op.py [hunk 6]
```diff
@@ -168,9 +230,67 @@ def fuser_backward(
...
+            grad_input = grad_output
+        return grad_input, grad_params, grad_extra_inputs
+
+    @classmethod
+    def forward_compute(cls, args: Any) -> tuple:
+        """Return (output, extra_outputs per basic op, fresh aux tensors)."""
+        raise NotImplementedError
+
+    @classmethod
+    def forward_fake(cls, args: Any) -> tuple:
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/op.py:250]: Nit: It may be helpful to indicate that this is supposed to imitate `forward_compute`:
```suggestion
    def forward_compute_fake(cls, args: Any) -> tuple:
```
Similar for `backward_compute_fake`.

File: transformer_engine/pytorch/ops/op.py [hunk 7]
```diff
@@ -59,6 +60,55 @@ def save_for_backward(self, *tensors: Optional[torch.Tensor]) -> None:
...
+            arg_type=cls.fwd_args_type,
+            impl=cls.forward_compute,
+            fake_impl=cls.forward_fake,
+        )
+        backward = register_custom_op(
+            op_name=f"{name}_backward",
+            arg_type=cls.bwd_args_type,
+            impl=cls.backward_compute,
+            fake_impl=cls.backward_fake,
+        )
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/op.py:91]: Noting that we assume that forward and backward ops are defined together. This makes sense for the case with a single op, but it may not hold for fused ops (a fused op might only have a forward impl, or only have a backward impl). Can we handle the case where the forward supports `torch.compile` and the backward needs a graph break, or vice versa?

**Review by timmoon10 — DISMISSED** [review]
Overall LGTM

File: transformer_engine/pytorch/ops/op.py [hunk 8]
```diff
@@ -59,6 +60,56 @@ def save_for_backward(self, *tensors: Optional[torch.Tensor]) -> None:
...
+            )
+        cls.compile_ops = (ops[0], ops[1])
+
+    def compile_unsupported_reason(self, mode: str) -> Optional[str]:
+        """Why this operation cannot run through its custom op, or ``None``."""
+        if self.compile_ops[("forward", "backward").index(mode)] is None:
+            return f"{self.__class__.__name__} without a custom op for {mode}"
+        basic_ops = self.basic_ops if self.is_fused_op else (self,)
+        for op in basic_ops:
+            for quantizer_mode in ("forward", "backward"):
```
Comment (kshitij12345) [inline transformer_engine/pytorch/ops/op.py:103]: I think this should be `op.num_quantizers(mode)` and should not iterate over ("forward", "backward"). 
Otherwise, for the case where forward is supported but backward isn't, this would return unsupported.

Comment (timmoon10) [reply transformer_engine/pytorch/ops/op.py:103]: I could see the argument either way. I'm not worried about safety since the logic is more cautious than necessary. However, it might be annoying for inference-only use-cases to configure unnecessary backward quantizers.

**Review by timmoon10 — APPROVED** [review]
